File: identity/src/test/java/com/android/identity/ConnectionMethodTest.java
/*
 * Copyright 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.identity;

import androidx.test.filters.SmallTest;

import org.junit.Assert;
import org.junit.Test;

import java.util.OptionalLong;
import java.util.UUID;

public class ConnectionMethodTest {

    @Test
    @SmallTest
    public void testConnectionMethodNfc() {
        ConnectionMethodNfc cm = new ConnectionMethodNfc(4096, 32768);
        ConnectionMethodNfc decoded = ConnectionMethodNfc.fromDeviceEngagement(cm.toDeviceEngagement());
        Assert.assertEquals(cm.getCommandDataFieldMaxLength(), decoded.getCommandDataFieldMaxLength());
        Assert.assertEquals(cm.getResponseDataFieldMaxLength(), decoded.getResponseDataFieldMaxLength());
        Assert.assertEquals("[\n" +
                "  1,\n" +
                "  1,\n" +
                "  {\n" +
                "    0 : 4096,\n" +
                "    1 : 32768\n" +
                "  }\n" +
                "]", Util.cborPrettyPrint(cm.toDeviceEngagement()));
    }

    @Test
    @SmallTest
    public void testConnectionMethodBle() {
        UUID uuidPeripheral = new UUID(0, 1);
        UUID uuidCentral = new UUID(123456789, 987654321);
        ConnectionMethodBle cm = new ConnectionMethodBle(
                true,
                true,
                uuidPeripheral,
                uuidCentral);
        ConnectionMethodBle decoded = ConnectionMethodBle.fromDeviceEngagement(cm.toDeviceEngagement());
        Assert.assertTrue(cm.getSupportsPeripheralServerMode());
        Assert.assertTrue(cm.getSupportsCentralClientMode());
        Assert.assertEquals(uuidPeripheral, cm.getPeripheralServerModeUuid());
        Assert.assertEquals(uuidCentral, cm.getCentralClientModeUuid());
        Assert.assertEquals("[\n" +
                "  2,\n" +
                "  1,\n" +
                "  {\n" +
                "    0 : true,\n" +
                "    1 : true,\n" +
                "    10 : [0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01],\n" +
                "    11 : [0x00, 0x00, 0x00, 0x00, 0x07, 0x5b, 0xcd, 0x15, 0x00, 0x00, 0x00, 0x00, 0x3a, 0xde, 0x68, 0xb1]\n" +
                "  }\n" +
                "]", Util.cborPrettyPrint(cm.toDeviceEngagement()));
    }

    @Test
    @SmallTest
    public void testConnectionMethodBleOnlyCentralClient() {
        UUID uuid = new UUID(123456789, 987654321);
        ConnectionMethodBle cm = new ConnectionMethodBle(
                false,
                true,
                null,
                uuid);
        ConnectionMethodBle decoded = ConnectionMethodBle.fromDeviceEngagement(cm.toDeviceEngagement());
        Assert.assertFalse(cm.getSupportsPeripheralServerMode());
        Assert.assertTrue(cm.getSupportsCentralClientMode());
        Assert.assertNull(cm.getPeripheralServerModeUuid());
        Assert.assertEquals(uuid, cm.getCentralClientModeUuid());
        Assert.assertEquals("[\n" +
                "  2,\n" +
                "  1,\n" +
                "  {\n" +
                "    0 : false,\n" +
                "    1 : true,\n" +
                "    11 : [0x00, 0x00, 0x00, 0x00, 0x07, 0x5b, 0xcd, 0x15, 0x00, 0x00, 0x00, 0x00, 0x3a, 0xde, 0x68, 0xb1]\n" +
                "  }\n" +
                "]", Util.cborPrettyPrint(cm.toDeviceEngagement()));
    }

    @Test
    @SmallTest
    public void testConnectionMethodBleOnlyPeripheralServer() {
        UUID uuid = new UUID(0, 1);
        ConnectionMethodBle cm = new ConnectionMethodBle(
                true,
                false,
                uuid,
                null);
        ConnectionMethodBle decoded = ConnectionMethodBle.fromDeviceEngagement(cm.toDeviceEngagement());
        Assert.assertTrue(cm.getSupportsPeripheralServerMode());
        Assert.assertFalse(cm.getSupportsCentralClientMode());
        Assert.assertEquals(uuid, cm.getPeripheralServerModeUuid());
        Assert.assertNull(cm.getCentralClientModeUuid());
        Assert.assertEquals("[\n" +
                "  2,\n" +
                "  1,\n" +
                "  {\n" +
                "    0 : true,\n" +
                "    1 : false,\n" +
                "    10 : [0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01]\n" +
                "  }\n" +
                "]", Util.cborPrettyPrint(cm.toDeviceEngagement()));
    }

    @Test
    @SmallTest
    public void testConnectionMethodWifiAware() {
        ConnectionMethodWifiAware cm = new ConnectionMethodWifiAware(
                "foobar",
                OptionalLong.of(42),
                OptionalLong.of(43),
                new byte[] {1, 2});
        ConnectionMethodWifiAware decoded = ConnectionMethodWifiAware.fromDeviceEngagement(cm.toDeviceEngagement());
        Assert.assertEquals("foobar", cm.getPassphraseInfoPassphrase());
        Assert.assertEquals(42, cm.getChannelInfoChannelNumber().getAsLong());
        Assert.assertEquals(43, cm.getChannelInfoOperatingClass().getAsLong());
        Assert.assertArrayEquals(new byte[] {1, 2}, cm.getBandInfoSupportedBands());
        Assert.assertEquals("[\n" +
                "  3,\n" +
                "  1,\n" +
                "  {\n" +
                "    0 : 'foobar',\n" +
                "    2 : 42,\n" +
                "    1 : 43,\n" +
                "    3 : [0x01, 0x02]\n" +
                "  }\n" +
                "]", Util.cborPrettyPrint(cm.toDeviceEngagement()));
    }

    @Test
    @SmallTest
    public void testConnectionMethodRestApi() {
        ConnectionMethodHttp cm = new ConnectionMethodHttp("https://www.example.com/mdocReader");
        ConnectionMethodHttp decoded = ConnectionMethodHttp.fromDeviceEngagement(cm.toDeviceEngagement());
        Assert.assertEquals(decoded.getUri(), cm.getUri());
        Assert.assertEquals("[\n" +
                "  4,\n" +
                "  1,\n" +
                "  {\n" +
                "    0 : 'https://www.example.com/mdocReader'\n" +
                "  }\n" +
                "]", Util.cborPrettyPrint(cm.toDeviceEngagement()));
    }
}

File: identity-mdoc/src/commonTest/kotlin/com/android/identity/mdoc/connectionmethod/ConnectionMethodTest.kt
/*
 * Copyright 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.identity.mdoc.connectionmethod

import com.android.identity.cbor.Cbor
import com.android.identity.cbor.DiagnosticOption
import com.android.identity.mdoc.transport.MdocTransport
import com.android.identity.util.UUID
import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class ConnectionMethodTest {
    @Test
    fun testConnectionMethodNfc() {
        val cm = ConnectionMethodNfc(4096, 32768)
        val decoded = ConnectionMethod.fromDeviceEngagement(cm.toDeviceEngagement()) as ConnectionMethodNfc?
        assertNotNull(decoded)
        assertEquals(decoded.commandDataFieldMaxLength, decoded.commandDataFieldMaxLength)
        assertEquals(decoded.responseDataFieldMaxLength, decoded.responseDataFieldMaxLength)
        assertEquals(
            """[
  1,
  1,
  {
    0: 4096,
    1: 32768
  }
]""", Cbor.toDiagnostics(cm.toDeviceEngagement(), setOf(DiagnosticOption.PRETTY_PRINT))
        )
        val ndefRecord = cm.toNdefRecord(listOf(), MdocTransport.Role.MDOC, false)!!.first
        val decodedNfc = ConnectionMethodNfc.fromNdefRecord(ndefRecord, MdocTransport.Role.MDOC)!!
        assertEquals(cm, decodedNfc)
    }

    @Test
    fun testConnectionMethodBle() {
        val uuidPeripheral = UUID(0U, 1U)
        val uuidCentral = UUID(123456789U, 987654321U)
        var cm = ConnectionMethodBle(
            true,
            true,
            uuidPeripheral,
            uuidCentral
        )
        val decoded = ConnectionMethod.fromDeviceEngagement(cm.toDeviceEngagement()) as ConnectionMethodBle?
        assertNotNull(decoded)
        assertTrue(decoded.supportsPeripheralServerMode)
        assertTrue(decoded.supportsCentralClientMode)
        assertEquals(uuidPeripheral, decoded.peripheralServerModeUuid)
        assertEquals(uuidCentral, decoded.centralClientModeUuid)
        assertEquals(
            """[
  2,
  1,
  {
    0: true,
    1: true,
    10: h'00000000000000000000000000000001',
    11: h'00000000075bcd15000000003ade68b1'
  }
]""", Cbor.toDiagnostics(cm.toDeviceEngagement(), setOf(DiagnosticOption.PRETTY_PRINT))
        )

        // For use in NFC, the UUIDs have to be the same
        val uuidBoth = UUID(0U, 2U)
        cm = ConnectionMethodBle(
            true,
            true,
            uuidBoth,
            uuidBoth
        )
        val ndefRecord = cm.toNdefRecord(listOf(), MdocTransport.Role.MDOC, false).first
        val decodedNfc = ConnectionMethodBle.fromNdefRecord(ndefRecord, MdocTransport.Role.MDOC, null)!!
        assertEquals(cm, decodedNfc)
    }

    @Test
    fun testConnectionMethodBleOnlyCentralClient() {
        val uuid = UUID(123456789U, 987654321U)
        val cm = ConnectionMethodBle(
            false,
            true,
            null,
            uuid
        )
        val decoded = ConnectionMethod.fromDeviceEngagement(cm.toDeviceEngagement()) as ConnectionMethodBle?
        assertNotNull(decoded)
        assertFalse(decoded.supportsPeripheralServerMode)
        assertTrue(decoded.supportsCentralClientMode)
        assertNull(decoded.peripheralServerModeUuid)
        assertEquals(uuid, decoded.centralClientModeUuid)
        assertEquals(
            """[
  2,
  1,
  {
    0: false,
    1: true,
    11: h'00000000075bcd15000000003ade68b1'
  }
]""", Cbor.toDiagnostics(cm.toDeviceEngagement(), setOf(DiagnosticOption.PRETTY_PRINT))
        )

        val ndefRecord = cm.toNdefRecord(listOf(), MdocTransport.Role.MDOC, false).first
        val decodedNfc = ConnectionMethodBle.fromNdefRecord(ndefRecord, MdocTransport.Role.MDOC, null)!!
        assertEquals(cm, decodedNfc)
    }

    @Test
    fun testConnectionMethodBleOnlyPeripheralServer() {
        val uuid = UUID(0U, 1U)
        val cm = ConnectionMethodBle(
            true,
            false,
            uuid,
            null
        )
        val decoded = ConnectionMethod.fromDeviceEngagement(cm.toDeviceEngagement()) as ConnectionMethodBle?
        assertNotNull(decoded)
        assertTrue(decoded.supportsPeripheralServerMode)
        assertFalse(decoded.supportsCentralClientMode)
        assertEquals(uuid, decoded.peripheralServerModeUuid)
        assertNull(decoded.centralClientModeUuid)
        assertEquals(
            """[
  2,
  1,
  {
    0: true,
    1: false,
    10: h'00000000000000000000000000000001'
  }
]""", Cbor.toDiagnostics(cm.toDeviceEngagement(), setOf(DiagnosticOption.PRETTY_PRINT))
        )

        val ndefRecord = cm.toNdefRecord(listOf(), MdocTransport.Role.MDOC, false).first
        val decodedNfc = ConnectionMethodBle.fromNdefRecord(ndefRecord, MdocTransport.Role.MDOC, null)!!
        assertEquals(cm, decodedNfc)
    }

    @Test
    fun testConnectionMethodWifiAware() {
        val cm = ConnectionMethodWifiAware(
            "foobar",
            42,
            43,
            byteArrayOf(1, 2)
        )
        val decoded = ConnectionMethod.fromDeviceEngagement(cm.toDeviceEngagement()) as ConnectionMethodWifiAware?
        assertNotNull(decoded)
        assertEquals("foobar", decoded!!.passphraseInfoPassphrase)
        assertEquals(42, decoded.channelInfoChannelNumber)
        assertEquals(43, decoded.channelInfoOperatingClass)
        assertContentEquals(byteArrayOf(1, 2), decoded.bandInfoSupportedBands)
        assertEquals(
            """[
  3,
  1,
  {
    0: "foobar",
    2: 42,
    1: 43,
    3: h'0102'
  }
]""", Cbor.toDiagnostics(cm.toDeviceEngagement(), setOf(DiagnosticOption.PRETTY_PRINT))
        )
    }

    @Test
    fun testConnectionMethodRestApi() {
        val cm = ConnectionMethodHttp("https://www.example.com/mdocReader")
        val decoded = ConnectionMethod.fromDeviceEngagement(cm.toDeviceEngagement()) as ConnectionMethodHttp?
        assertNotNull(decoded)
        assertEquals(decoded.uri, cm.uri)
        assertEquals(
            """[
  4,
  1,
  {
    0: "https://www.example.com/mdocReader"
  }
]""", Cbor.toDiagnostics(cm.toDeviceEngagement(), setOf(DiagnosticOption.PRETTY_PRINT))
        )
    }

    @Test
    fun testConnectionMethodDisambiguate() {
        var ble: ConnectionMethodBle
        val disambiguated = ConnectionMethod.disambiguate(
            listOf(
                ConnectionMethodHttp("https://www.example.com/mdocReader"),
                ConnectionMethodBle(
                    true,
                    true,
                    UUID(0U, 1U),
                    UUID(0U, 2U)
                )
            )
        )
        assertEquals(3, disambiguated.size.toLong())
        assertTrue(disambiguated[0] is ConnectionMethodHttp)
        assertTrue(disambiguated[1] is ConnectionMethodBle)
        assertTrue(disambiguated[2] is ConnectionMethodBle)
        ble = disambiguated[1] as ConnectionMethodBle
        assertFalse(ble.supportsPeripheralServerMode)
        assertTrue(ble.supportsCentralClientMode)
        assertNull(ble.peripheralServerModeUuid)
        assertEquals(UUID(0U, 2U), ble.centralClientModeUuid)
        ble = disambiguated[2] as ConnectionMethodBle
        assertTrue(ble.supportsPeripheralServerMode)
        assertFalse(ble.supportsCentralClientMode)
        assertEquals(UUID(0U, 1U), ble.peripheralServerModeUuid)
        assertNull(ble.centralClientModeUuid)
    }

    @Test
    fun testConnectionMethodCombineUuidNotSame() {
        val disambiguated = listOf(
            ConnectionMethodHttp("https://www.example.com/mdocReader"),
            ConnectionMethodBle(
                true,
                false,
                UUID(0U, 1U),
                null
            ),
            ConnectionMethodBle(
                false,
                true,
                null,
                UUID(0U, 2U)
            )
        )
        assertFailsWith<IllegalArgumentException> {
            val combined = ConnectionMethod.combine(disambiguated)
        }
    }

    @Test
    fun testConnectionMethodCombineUuid() {
        val uuid = UUID(0U, 3U)
        val disambiguated = listOf(
            ConnectionMethodHttp("https://www.example.com/mdocReader"),
            ConnectionMethodBle(
                true,
                false,
                uuid,
                null
            ),
            ConnectionMethodBle(
                false,
                true,
                null,
                uuid
            )
        )
        val combined = ConnectionMethod.combine(disambiguated)
        assertEquals(2, combined.size.toLong())
        val ble = combined[0] as ConnectionMethodBle
        assertTrue(ble.supportsPeripheralServerMode)
        assertTrue(ble.supportsCentralClientMode)
        assertEquals(uuid, ble.peripheralServerModeUuid)
        assertEquals(uuid, ble.centralClientModeUuid)
    }
}